nexus/src/authn/saga.rs

```diff
@@ -0,0 +1,31 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Serialization of authentication context for sagas
+// TODO-security: This is not fully baked.  Right now, we serialize the actor
+// id.  We should think through that a little more.  Should we instead preload a
+// bunch of roles and then serialize that, for example?
+
+use crate::authn;
+use crate::context::OpContext;
+use serde::Deserialize;
+use serde::Serialize;
+
+/// Serialized form of an `OpContext`
+// NOTE: This structure must be versioned carefully.  (That's true of all saga
+// structures, but this one has a particularly large blast radius.)
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Serialized {
+    kind: authn::Kind,
+}
```

Comment:
What was the rationale behind creating a new struct instead of:
- Allowing a caller to get `authn::Kind` from an `authn::Context`
- Implementing `From<authn::Kind>` for `authn::Context`?

(I know the comment says `OpContext`, but this seems much more specifically targeting `authn::Context`)

That's basically what this is  right?  Something which should stay 1:1 with the "current authentication information", or some scheme to re-acquire it? Since `Kind` includes "all the possible ways the user may be authenticated", that seems like a reasonable proxy.

Either way - could we change some of the names being used here?

- I know this is intended to be used *by* sagas, but this data structure had nothing directly to do with sagas. If a caller wanted to record "who authenticated an operation at a point-in-time", they *could* use this code, but that seems like it would be a divergence from the name.
- If we stick with a separate struct - `Serialized` seems a bit generic, no? Tons of structures which are serializable exist in the `authn` module.

Reply:
I agree with your feeling that there's something off about the structure and name here.  Let me address your specific points first:

> What was the rationale behind creating a new struct instead of:
> 
>     * Allowing a caller to get `authn::Kind` from an `authn::Context`
> 
>     * Implementing `From<authn::Kind>` for `authn::Context`?
...
> (I know the comment says `OpContext`, but this seems much more specifically targeting `authn::Context`)
> 
> That's basically what this is right? Something which should stay 1:1 with the "current authentication information", or some scheme to re-acquire it? Since `Kind` includes "all the possible ways the user may be authenticated", that seems like a reasonable proxy.

I've been trying to keep the internal details of `authn::Context` as private as possible because I expect them to change as we flesh out authn.  What you said is currently true.  I don't know if that will be true going forward.

> Either way - could we change some of the names being used here?
> 
>     * I know this is intended to be used _by_ sagas, but this data structure had nothing directly to do with sagas. If a caller wanted to record "who authenticated an operation at a point-in-time", they _could_ use this code, but that seems like it would be a divergence from the name.

There are different reasons that someone might want a serialized idea of "who authenticated an operation".  These use cases might have very different security properties.  You might want this information for a debug log, a formal audit log, or to do an authz check in the context of the request, or to propagate the authentication to some other context, possibly at a later time (as in the case of a saga action, but could also be true if we passed this information in a request to some other control plane service).  Some of these representations could involve actual secrets (e.g., a bearer token), while others absolutely should never include a secret (the debug log entry).  This change creates one representation that's (barely) suitable for one of these purposes.  As you said it's simple enough to be used for many of them today.  But it seems unlikely that will be true in the future.

Now, we could go either way on this.  If we expect to have multiple representations in the future, we could label this structure as general-purpose today and trust that our future selves will be careful enough to recognize when security demands a different representation and build new representations as needed.  Or we could label it today as "suitable for use in this specific case only" and trust our future selves, with a second use case in hand, to decide whether it's suitable and refactor as needed.  I generally tend towards the latter -- I don't like to describe something as generic when I only really know about one use case.

>     * If we stick with a separate struct - `Serialized` seems a bit generic, no? Tons of structures which are serializable exist in the `authn` module.

Yeah, as we've previously discussed, I'd been trying to let the module name describe it.  It's `authn::saga::Serialized` -- a serialized authn [context] for use in sagas.  `authn::saga::SerializedIdentity`?  `authn::SerializedIdentityForSaga`? `authn::SagaIdentity`?

--- 

So like I said, I agree with you that there's something funny about the structure or naming here.  But I'm not sure how better to phrase it.  To restate my goals:

- have a type that can be easily incorporated directly into a saga's parameters
- have it be very easy to get that type from an `OpContext` and to reconstruct an `OpContext` from it [along with other information that's also part of the OpContext]
- so far, that type could be `authn::Context` directly or as you said `authn::Kind`.  That sort of ties the in-memory form (which is mainly intended for the use case of authorizing an operation in the context of this request) to the serialized form (which is intended for storing to disk and reusing at a later time in some other context).
- I also want to make it hard for someone to (accidentally) serialize the internal implementation details of an `authn::Context` to somewhere else (e.g., the public API or some other database record or a log file) without considering the consequences -- i.e., if somebody wants to do that, I want them to ask "hmm, maybe `authn::Context` [or `authn::Kind`] doesn't impl Serialize for a reason?"